src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1175 @@
+
+=item
+
+LINE_VALUE_BOOLEAN: interpret the value as a boolean rendered as C<yes> or C<no>
+
+=item
+
+LINE_VALUE_ARRAY: the value is an array. Rendering controlled by LINE_VALUE_OPT_xxx constants.
+
+=item
+
+LINE_VALUE_HASH_KEYS: the value is hash whose keys are the value. Rendering similar to arrays.
+
+=item
+
+LINE_VALUE_STRING_HASH: the value is a hash of string. Rendering controlled by LINE_VALUE_OPT_xxx constants.
+
+=item
+
+LINE_VALUE_INSTANCE_PARAMS: specific to L<ncm-xrootd>
+
+=back
+
+=cut
+
+use enum qw(
+  LINE_VALUE_AS_IS
+  LINE_VALUE_BOOLEAN
+  LINE_VALUE_ARRAY
+  LINE_VALUE_HASH_KEYS
+  LINE_VALUE_STRING_HASH
+  LINE_VALUE_INSTANCE_PARAMS
+  );
+
+=pod
+
+=head3 LINE_VALUE_OPT_xxx: options for rendering the value
+
+These options mainly apply to lists and hashes and are interpreted as a bitmask.
+
+=over
+
+=item
+
+LINE_VALUE_OPT_SINGLE: each value must be a separate instance of the keyword (multiple lines)
+
+=item
+
+LINE_VALUE_OPT_UNIQUE: each values are concataneted as a space-separated string
+
+=item
+
+LINE_VALUE_OPT_SORTED: values are sorted
+
+=back
+
+=cut
+
+use enum qw(
+  BITMASK: LINE_VALUE_OPT_SINGLE
+  LINE_VALUE_OPT_UNIQUE
+  LINE_VALUE_OPT_SORTED
+  );
+
+# Internal constants
+Readonly my $LINE_FORMAT_DEFAULT            => LINE_FORMAT_SH_VAR;
+Readonly my $LINE_QUATTOR_COMMENT           => "\t\t# Line generated by Quattor";
+Readonly my $LINE_OPT_DEF_REMOVE_IF_UNDEF   => 0;
+Readonly my $LINE_OPT_DEF_ALWAYS_RULES_ONLY => 0;
+Readonly my $RULE_CONDITION_ALWAYS          => 'ALWAYS';
+Readonly my $RULE_OPTION_SET_GLOBAL         => 'GLOBAL';
+
+
+# Export constants used to build rules
+# Needs to be updated when a constant is added or removed
+Readonly my @RULE_CONSTANTS => qw(
+  LINE_FORMAT_SH_VAR
+  LINE_FORMAT_ENV_VAR
+  LINE_FORMAT_KEY_VAL
+  LINE_FORMAT_KEY_VAL_SETENV
+  LINE_FORMAT_KEY_VAL_SET
+  LINE_VALUE_AS_IS
+  LINE_VALUE_BOOLEAN
+  LINE_VALUE_INSTANCE_PARAMS
+  LINE_VALUE_ARRAY
+  LINE_VALUE_HASH_KEYS
+  LINE_VALUE_STRING_HASH
+  LINE_VALUE_OPT_SINGLE
+  LINE_VALUE_OPT_UNIQUE
+  LINE_VALUE_OPT_SORTED
+  );
+
+
+our @EXPORT_OK;
+our %EXPORT_TAGS;
+push @EXPORT_OK, @RULE_CONSTANTS;
+$EXPORT_TAGS{rule_constants} = \@RULE_CONSTANTS;
+
+
+# Backup file extension
+Readonly my $BACKUP_FILE_EXT => ".old";
+
+=pod
+
+=head2 Public methods
+
+=over
+
+=item updateFile
+
+Update configuration file contents,  applying configuration rules.
+
+Arguments :
+    config_rules: config rules corresponding to the file to build
+    config_options: configuration parameters used to build actual configuration
+    options: a hash setting options to modify the behaviour of this function
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove matching configuration line is rule condition is not met (D: false)
+
+Return value
+    sucess: 1
+    argument error: undef
+    
```

Comment:
why only undef in case of argument error? what about invalid rules etc? those should also return undef, no?